### PR TITLE
Slice 2.0 - only byte[] as backend

### PR DESCRIPTION
**Pull request by dain**
This proposes that for Slice 2.0 we drop support for a slice backed by anything other than byte[].  Currently, there are very few usages for Slices for over other array types, and even then they are almost exclusively used for the bulk transfer between a byte[] Slice and other primitive arrays.  There are no known uses of Slice over direct memory, as most systems use ByteBuffer for compatibility with NIO.  In the future the Java Memory APIs will be a better approach to native memory access.

Changing Slice to only support byte[] drastically simplifies the codebase, and more importantly makes it easier for users.  A Slice can not be unconditionally unwrapped to a byte array, and the library can provide accessors without bounds checks without worry of JVM crashes.  Finally, the library may be faster in complex environments because there is only one possible backing memory type, so all methods become monomorphic (this is a purely speculative assumption).

### CODE REVIEW

File: src/main/java/io/airlift/slice/Slice.java [hunk 1]
```diff
@@ -156,133 +110,22 @@ public final class Slice
...
     }
 
     /**
      * Creates a slice for directly accessing the base object.
      */
-    Slice(@Nullable Object base, long address, int size, long retainedSize, @Nullable Object reference)
+    Slice(byte[] base, long address, int size, long retainedSize)
     {
-        if (address <= 0) {
+        if (address < ARRAY_BYTE_BASE_OFFSET) {
```
Comment (electrum) [inline src/main/java/io/airlift/slice/Slice.java]: We can also validate that `size` is within the bounds of the array

File: src/main/java/io/airlift/slice/Slice.java
Comment (electrum) [inline src/main/java/io/airlift/slice/Slice.java:1]: Typo in commit message "getters to without bounds check"

Comment (electrum) [reply src/main/java/io/airlift/slice/Slice.java:1]: Why are these needed at all? The Javadoc for `MethodHandles.byteArrayViewVarHandle()` says

> Access of bytes at a given index will result in an ArrayIndexOutOfBoundsException if the index is less than 0 or greater than the byte[] array length minus the size (in bytes) of T.

So it seems like we could remove all the explicit bounds checks in Slice.

**Review by electrum — APPROVED** [review]

File: src/main/java/io/airlift/slice/Slice.java
Comment (dain) [reply src/main/java/io/airlift/slice/Slice.java:1]: The normal getters need check to ensure that only the visible slice of the byte array is accessed.  The unchecked getters will, for example, allow negative offsets for a slice that is inset in the byte[]

File: src/main/java/io/airlift/slice/Slice.java (hunk 1 above)
Comment (dain) [reply src/main/java/io/airlift/slice/Slice.java]: This is added when the code is switched to byte[]

**Pull request merged**